### PR TITLE
implement sac for discrete action settings

### LINKED ISSUE #71: Is there any example for SAC for discrete
Hi ,i search over all the project ,but i have not find any example for SAC discrete, how can i build it?or could you please add an example for SAC discrete

### PR DESCRIPTION
**Pull request by danagi**
- [ ] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [x] algorithm implementation fix
    + [ ] documentation modification
    + [x] new feature
- [ ] If applicable, I have mentioned the relevant/related issue(s)

Less important but also useful:

- [ ] I have visited the [source website](https://github.com/thu-ml/tianshou)
- [ ] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, sys
  print(tianshou.__version__, torch.__version__, sys.version, sys.platform)
  ```

### CODE REVIEW

**Review by Trinkle23897 — DISMISSED** [review]

File: tianshou/policy/modelfree/discrete_sac.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                         **kwargs)
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        obs = getattr(batch, input)
```
Comment (duburcqa) [inline tianshou/policy/modelfree/discrete_sac.py]: `getattr` is necessary ?

File: tianshou/policy/modelfree/discrete_sac.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        dist = Categorical(logits=logits)
+        act = dist.sample()
+        return Batch(logits=logits, act=act, state=h, dist=dist)
+
+    def _target_q(
+        self, buffer: ReplayBuffer, indice: np.ndarray
+    ) -> torch.Tensor:
+        batch = buffer[indice]  # batch.obs: s_{t+n}
+        with torch.no_grad():
+            obs_next_result = self(batch, input="obs_next")
```
Comment (duburcqa) [inline tianshou/policy/modelfree/discrete_sac.py:82]: I would rather use `self.forward` since calling `self` directly in internal methods is usually.

File: tianshou/policy/modelfree/discrete_sac.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            )
+            target_q = target_q.sum(dim=-1) + self._alpha * dist.entropy()
+        return target_q
+
+    def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
+        weight = batch.pop("weight", 1.0)
+        target_q = batch.returns.flatten()
+        act = to_torch(batch.act, device=target_q.device, dtype=torch.long)
+        act = act.unsqueeze(1)
+        # critic 1
```
Comment (duburcqa) [inline tianshou/policy/modelfree/discrete_sac.py:97]: Odd spacing.

File: tianshou/policy/modelfree/discrete_sac.py (hunk 1 above)
Comment (Trinkle23897) [reply tianshou/policy/modelfree/discrete_sac.py]: Yes it is. `input` may be `obs_next`

Comment (duburcqa) [reply tianshou/policy/modelfree/discrete_sac.py]: Ok I see.

Comment (duburcqa) [reply tianshou/policy/modelfree/discrete_sac.py]: So why not taking advantage of dict-like key access, doing `batch[input]` ?

Comment (Trinkle23897) [reply tianshou/policy/modelfree/discrete_sac.py]: That's cool!

File: tianshou/policy/modelfree/discrete_sac.py (hunk 2 above)
Comment (danagi) [reply tianshou/policy/modelfree/discrete_sac.py:82]: It's not recommended to use `self.forward` directly because the internal `__call__` function manages all registered hooks.  Calling `self.forward` directly would skip those hooks. [https://discuss.pytorch.org/t/any-different-between-model-input-and-model-forward-input/3690/2](url)

Comment (duburcqa) [reply tianshou/policy/modelfree/discrete_sac.py:82]: I didn't know there was any difference between `__call__` and `forward`. Ok then !

**Review by duburcqa — APPROVED** [review]
